File: Data Structures/binarySearchTree.py
"""To convert a Binary Tree to Binary Search Tree."""
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def storeInOrder(root, ordered):
    """To store in order the data from the nodes of a given tree."""
    if root is None:
        return

    storeInOrder(root.left, ordered)
    ordered.append(root.data)
    storeInOrder(root.right, ordered)

def countNodes(root):
    """Count the number of nodes in a given tree."""
    if root is None:
        return
    
    if root.right==None:
        rightNode = 0
    else:
        rightNode = countNodes(root.right)

    if root.left==None:
        leftNode = 0
    else:
        leftNode = countNodes(root.left)

    numberOfNodes = rightNode + leftNode + 1

    return numberOfNodes

def arrayToBinarySearchTree(arr, root):
    """Create a Binary Search Tree from an array."""
    if root is None:
        return

    arrayToBinarySearchTree(arr, root.left)

    root.data = arr[0]
    arr.pop(0)

    arrayToBinarySearchTree(arr, root.right)

def binaryTreeToBinarySearchTree(root):
    """Converts a Binary Tree into a Binary Search Tree."""
    if root is None:
        return

    n = countNodes(root)
    arr = []
    storeInOrder(root, arr)
    arr.sort()
    arrayToBinarySearchTree(arr, root)

def printInOrder(root):
    """Print the data contained in the Tree."""
    if root is None:
        return
    printInOrder(root.left)
    print(root.data)
    printInOrder(root.right)

if __name__ == "__main__":
    root = Node(10)
    root.left = Node(30)
    root.right = Node(15)
    root.left.left = Node(23)
    root.right.right = Node(5)

    binaryTreeToBinarySearchTree(root)
    print("Data contained in the tree in order.\n")
    printInOrder(root)

File: Data Structures/binarySearchTree.c
/* Binary Tree to Binary Search Tree */
#include <stdio.h>
#include <stdlib.h>

// Node of a binary tree
struct node{
	int data;
	struct node* left;
	struct node* right;
};

// Helper functions
void storeInOrder(struct node* node, int ordered[], int* index){
	if (node == NULL)
		return;
	storeInOrder(node->left, ordered, index);

	ordered[*index] = node->data;

	(*index)++;

	storeInOrder(node->right, ordered, index);
}

int countNodes(struct node* root){
	if (root = NULL)
		return 0;

	return countNodes(root->left) + countNodes(root->right) + 1;
}

int compare(const void* a, const void* b){
	return (*(int*)a - *(int*)b);
}

void fromArrayToBinarySearchTree(int* arr, struct node* root, int* index){
	if (root = NULL)
		return;

	fromArrayToBinarySearchTree(arr, root->left, index);

	root->data = arr[*index];
	(*index)++;

	fromArrayToBinarySearchTree(arr, root->right, index);
}

// Convert a Binary Tree to a Binary Search Tree
void binaryTreeToBinarySearchTree(struct node* root){
	// base case
	if (root == NULL)
		return;

	int n = countNodes(root);
	// pass the element to an array
	int arr[n];
	int i = 0;
	storeInOrder(root, arr, &i);
	// sort them
	qsort(arr, n, sizeof(arr[0]), compare);
	// copy back
	i = 0;
	fromArrayToBinarySearchTree(arr, root, &i);
	// free the memory
	free(arr);
}

// New node for the Binary Tree
struct node* newNode(int data){
	struct node* temp = (struct node*)malloc(sizeof(struct node)); //malloc(sizeof(struct node)); //(struct node*)malloc(sizeof(struct node));
	temp->data = data;
	temp->left = NULL;
	temp->right = NULL;
	return temp;
}

// helper
void printInOrder(struct node* node){
	if(node==NULL)
		return;
	printInOrder(node->left);
	printf("%d ", node->data);
	printInOrder(node->right);
}


// Driver function
int main(){
	struct node* root = NULL;
	root=newNode(10);
	root->left=newNode(30);
	root->right=newNode(15);
	root->left->left=newNode(20);
	root->right->right=newNode(5);

	binaryTreeToBinarySearchTree(root);

	printf("Printed in order");

	printInOrder(root);

	return 0;
}
